mtp_api/apps/transaction/views.py

```diff
@@ -52,9 +54,87 @@ class TransactionView(
 ):
     queryset = Transaction.objects.all()
     serializer_class = TransactionSerializer
+    patch_serializer_class = CreditedOnlyTransactionSerializer
     filter_backends = (
         filters.DjangoFilterBackend,
         filters.OrderingFilter,
     )
-    filter_class = TransactionFilter
+
+    filter_class = TransactionStatusFilter
+
     ordering = ('received_at',)
+    permission_classes = (IsAuthenticated, IsOwnPrison,)
+
+    def get_queryset(self, filter_by_user=True, filter_by_prison=True):
+        qs = super(TransactionView, self).get_queryset()
+
+        prison_id = self.kwargs.get('prison_id')
+        user_id = self.kwargs.get('user_id')
+
+        if prison_id and filter_by_prison:
+            qs = qs.filter(prison_id=prison_id)
+
+        if user_id and filter_by_user:
+            qs = qs.filter(owner__id=user_id)
+
+        return qs
+
+    def take(self, request, *args, **kwargs):
+        self.permission_classes = list(self.permission_classes) + [IsOwner]
+        self.check_permissions(request)
+
+        slice_size = self.get_slice_limit(request)
+
+        with transaction.atomic():
+            pending = self.get_queryset(filter_by_user=False).available().select_for_update()
+            slice_pks = pending.values_list('pk', flat=True)[:slice_size]
+
+            queryset = self.get_queryset(filter_by_user=False).filter(pk__in=slice_pks)
+            queryset.update(owner=request.user)
```

Comment:
Doesn't this clash with the route? If I understand correctly, the route is `<prison_id>/<user_id>/take'`, but we're passing `request.user`… Doesn't that mean I could hit `<prison_id>/<someone_else>/take` and it would assign to me? What's the expected behaviour? If we're just using `request.user`, I'd suggest we make the route `<prison_id>/take`.